pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go

```diff
@@ -2,35 +2,106 @@ package remote
+type remoteAlertmanager interface {
+	notifier.Alertmanager
+	CompareAndSendConfiguration(context.Context, *models.AlertConfiguration) error
+	CompareAndSendState(context.Context) error
+}
+
 type RemoteSecondaryForkedAlertmanager struct {
 	log log.Logger
 
 	internal notifier.Alertmanager
-	remote   notifier.Alertmanager
+	remote   remoteAlertmanager
+
+	lastSync     time.Time
+	syncInterval time.Duration
 }
 
-func NewRemoteSecondaryForkedAlertmanager(l log.Logger, internal, remote notifier.Alertmanager) *RemoteSecondaryForkedAlertmanager {
-	return &RemoteSecondaryForkedAlertmanager{
-		log:      l,
-		internal: internal,
-		remote:   remote,
+type RemoteSecondaryConfig struct {
+	// SyncInterval determines how often we should attempt to synchronize
+	// state and configuration on the external Alertmanager.
+	SyncInterval time.Duration
+	Logger       log.Logger
+}
+
+func (c *RemoteSecondaryConfig) Validate() error {
+	if c.Logger == nil {
+		return fmt.Errorf("logger cannot be nil")
 	}
+	return nil
+}
+
+func NewRemoteSecondaryForkedAlertmanager(cfg RemoteSecondaryConfig, internal notifier.Alertmanager, remote remoteAlertmanager) (*RemoteSecondaryForkedAlertmanager, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+	return &RemoteSecondaryForkedAlertmanager{
+		log:          cfg.Logger,
+		internal:     internal,
+		remote:       remote,
+		syncInterval: cfg.SyncInterval,
+	}, nil
 }
 
 // ApplyConfig will only log errors for the remote Alertmanager and ensure we delegate the call to the internal Alertmanager.
 // We don't care about errors in the remote Alertmanager in remote secondary mode.
 func (fam *RemoteSecondaryForkedAlertmanager) ApplyConfig(ctx context.Context, config *models.AlertConfiguration) error {
-	if err := fam.remote.ApplyConfig(ctx, config); err != nil {
-		fam.log.Error("Error applying config to the remote Alertmanager", "err", err)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	// Figure out if we need to sync the external Alertmanager in another goroutine.
+	go func() {
+		defer wg.Done()
+		// If the Alertmanager has not been marked as "ready" yet, delegate the call to the remote Alertmanager.
+		// This will perform a readiness check and sync the Alertmanagers.
+		if !fam.remote.Ready() {
+			if err := fam.remote.ApplyConfig(ctx, config); err != nil {
+				fam.log.Error("Error applying config to the remote Alertmanager", "err", err)
+				return
+			}
+			fam.lastSync = time.Now()
+			return
+		}
+
+		// If the Alertmanager was marked as ready but the sync interval has elapsed, sync the Alertmanagers.
+		if time.Since(fam.lastSync) >= fam.syncInterval {
+			fam.syncStateAndConfig(ctx, config)
+		}
+	}()
+
+	// Call ApplyConfig on the internal Alertmanager - we only care about errors for this one.
+	err := fam.internal.ApplyConfig(ctx, config)
+	wg.Wait()
+	return err
+}
+
+func (fam *RemoteSecondaryForkedAlertmanager) syncStateAndConfig(ctx context.Context, config *models.AlertConfiguration) {
```

Comment:
I don't think this indirection is doing us any favours, I'd consider removed it.